keps/sig-storage/1790-recover-resize-failure/README.md

```diff
@@ -103,73 +106,139 @@ reduces the PVC request size, for both CSI and in-tree plugins they are designed
+
+When user reduces `pvc.Spec.Resources`, expansion-controller will set `pvc.Status.AllocatedResources` to lower value only if:
+1. If `pvc.Status.ResizeStatus` is `ResizeFailedOnController` (indicating that previous expansion to last known `allocatedResources` failed with a final error).
+2. If `pvc.Status.ResizeStatus` is `ResizeFailedOnNode` and SP supports node-only expansion (indicating that previous expansion to last known `allocatedResources` failed on node with a final error).
+
+One key exception is - `pvc.Status.Allocatedresources` will not be lowered if expansion succeeded on the controller but failed on the node with a final error. The implementation of new `allocatedResources` will look like:
+
+
+```golang
+func (ctrl *resizeController) getNewSize(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) resource.Quantity {
+    newSize := pvc.Spec.Resources.Requests.Storage()
+    if pvc.Status.ResizeStatus == nil || *pvc.Status.ResizeStatus == "" {
+        return *newSize
+    }
+    resizeStatus := *pvc.Status.ResizeStatus
+    allocatedSize := pvc.Status.AllocatedResources.Storage()
+
+    switch resizeStatus {
+    case v1.PersistentVolumeClaimResizeInProgress:
+        // If resize was inprogress before this means - we should keep trying to finish
+        // previously started expansion to last known `allocatedSize`.
+        if allocatedSize != nil {
+            newSize = allocatedSize
+        }
+    case v1.PersistentVolumeClaimResizeFailedOnController:
+        // if expansion failed on controller with a final error, then we can safely use new user requested size
+        newSize = pvc.Spec.Resources.Requests.Storage()
+    case v1.PersistentVolumeClaimResizeFailedOnNode:
+        if newSize.Cmp(*allocatedSize) > 0 {
+            newSize = pvc.Spec.Resources.Requests.Storage()
+        } else {
+            // newSize is smaller than whatever we tried to expand before
+            // and if expansion failed on the node, but it succeeded in control-plane
+            // then it is not safe to use lower value from pvc.Spec
+            if ctrl.resizer.SupportsControllerExpansion() {
+                newSize = allocatedSize
+            }
+        }
+
+    }
+    return *newSize
+}
+```
+
+One more key design choice is - once expansion on node failed and user requested lower size (by modifying `pvc.Spec.Resources`) - kubelet must not retry expanding the same volume to previous `allocatedResources`, because external-resizer will consider `ResizeFailedOnNode` state as a terminal error and allow lowering of `allocatedResources` - and at the same time kubelet will try to expand the volume to higher `allocatedResources`. This will be solved in kubelet by following code:
```

Comment:
Discussed this offline with @jsafrane and he correclty pointed out that - this check here is not good enough. There is still a chance that while kubelet is increasing the volume size, user may reduce the size and resize-controller may pick lower value (because PVC still has `ResizeFailedOnNode` condition).

To fix - we are going to need additional `ResizeStatus` value. I will update the KEP.

Reply:
We managed to solve this problem with new design.